aggregate_root/spec/snapshot_repository_spec.rb

```diff
@@ -0,0 +1,164 @@
+          @status = :created
+        end
+
+        def apply_order_expired(_)
+          @status = :expired
+        end
+
+        def apply_order_canceled(_)
+          @status = :canceled
+        end
+      end
+
+      Order
+    end
+
+    specify 'initialization' do
+      expect { AggregateRoot::SnapshotRepository.new(event_store, 0) }
+        .to raise_error(ArgumentError, "interval must be greater than 0")
+    end
+
+    specify 'storing snapshot' do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store)
+      allow(event_store).to receive(:publish)
+
+      order.apply(order_created)
+      repository.store(order, stream_name)
+      expect(event_store).to have_received(:publish).with(
+        [order_created],
+        stream_name: stream_name,
+        expected_version: -1
+      )
+      expect_snapshot(order_created.event_id, 0, Marshal.dump(order))
+    end
+
+    specify 'storing snapshot with given interval' do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store, 2)
+      allow(event_store).to receive(:publish)
+
+      order.apply(order_created)
+      repository.store(order, stream_name)
+      expect_no_snapshot
+
+      order.apply(order_expired)
+      repository.store(order, stream_name)
+      expect_snapshot(order_expired.event_id, 1, Marshal.dump(order))
+    end
+
+    specify "restoring snapshot" do
+      order = order_klass.new(uuid)
+      repository = AggregateRoot::SnapshotRepository.new(event_store)
+
+      apply_and_publish(order, order_created, order_canceled)
+      expect_n_records_read(2) do
+        order_from_snapshot = repository.load(order_klass.new(uuid), stream_name)
+        expect(order_from_snapshot.status).to eq(:canceled)
+        expect(order_from_snapshot.version).to eq(1)
+      end
+
+      publish_snapshot(order, order_canceled.event_id, 1)
+      expect_n_records_read(1) do
+        order_from_snapshot = repository.load(order_klass.new(uuid), stream_name)
+        expect(order_from_snapshot.status).to eq(:canceled)
+        expect(order_from_snapshot.version).to eq(1)
+      end
+
+      apply_and_publish(order, order_expired)
+      expect_n_records_read(2) do
+        order_from_snapshot = repository.load(order_klass.new(uuid), stream_name)
+        expect(order_from_snapshot.status).to eq(:expired)
+        expect(order_from_snapshot.version).to eq(2)
+      end
+    end
+
+    private
+
+    def expect_n_records_read(n, &block)
+      reporting_repository.reset_records_read
+      yield block
+      expect(reporting_repository.records_read).to eq(n)
+    end
+
+    def apply_and_publish(order, *events)
+      events.each do |e|
+        order.apply(e)
+        event_store.publish(e, stream_name: stream_name)
+      end
+    end
+
+    def publish_snapshot(order, last_event_id, version)
+      event_store.publish(
+        AggregateRoot::SnapshotRepository::Snapshot.new(
```

Comment:
Works but seems brittle to changes in the aggregate itself. Shouldn't we expect a snapshot being recorded after interval and assume this is unecessary detail?